staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -212,13 +218,21 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	var unknown runtime.Unknown
 	internalEvent := &metav1.InternalEvent{}
 	outEvent := &metav1.WatchEvent{}
+	var bookmarkEvent *metav1.WatchEvent
 	buf := &bytes.Buffer{}
 	ch := s.Watching.ResultChan()
 	for {
 		select {
 		case <-cn.CloseNotify():
 			return
 		case <-timeoutCh:
+			klog.V(5).Infof("request allow bookmark: %v", s.AllowBookmark)
```

Comment:
The timeout mechanism may exist at the server and the client, so we will send a little time before the timeout instead of sending when the server times out. Simultaneously the original plan is to dynamically sending bookmarks periodically period based on the load of the server.